### PR TITLE
fix: [CDS-75505]: Exclude some paths in diff algorithm - QA

### PR DESCRIPTION
**Pull request by srdjan-harness**
### Summary

In the pipeline (when we use pipeline/stage template) for some entities like service when we chose service we are getting additional fields that are set as runtime inputs e.g. tags.
In this case diff-algorithm reporting that pipeline inputs and template inputs are different (in other words this means that template or pipeline(yaml) has been changed).

This pr exclude some paths (entities) for this and other edge cases.

https://harness.atlassian.net/browse/CDS-75505

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by srdjan-harness** [comment]
retrigger titlecheck

**Review by PratyushGar — APPROVED** [review]

**Review by bhavya-sinha — APPROVED** [review]

File: src/modules/70-pipeline/utils/pipelineReconcileUtils.ts [hunk 1]
```diff
@@ -42,29 +54,29 @@ function hasDiffTemplateInputsAndStageOrPipeline<T extends { [key: string]: any
...
   // To prevent this, if yaml is changed (isUpdated) we are not checking if there are more fileds in template than in pipeline.
   // This means that in some edge cases algorithm will not work as expected.
   if (!isUpdated) {
-    templatePaths_WithoutIndex_RuntimeOnly.forEach((tlPath: string) => {
-      if (pipelinePaths_WithoutIndex.indexOf(tlPath) === -1) {
+    templatePaths_RuntimeOnly.forEach((tlPath: string) => {
+      if (pipelinePaths.indexOf(tlPath) === -1) {
         hasDifference = true
       }
     })
```
Comment (knagurski) [inline src/modules/70-pipeline/utils/pipelineReconcileUtils.ts:74]: Not a problem with a hotfix, but a better way than `forEach` and `indexOf` would be `some` and `includes`.
```suggestion
    if (templatePaths_RuntimeOnly.some((tlPath: string) => pipelinePaths.includes(tlPath))) {
      hasDifference = true
    }
```

File: src/modules/70-pipeline/utils/pipelineReconcileUtils.ts [hunk 2]
```diff
@@ -75,7 +87,16 @@ function hasDiffTemplateInputsAndStageOrPipeline<T extends { [key: string]: any
...
+    return `${get(obj, fullPath)}`.startsWith('<+input>') && !excludePath(fullPath)
+  })
+}
+
+function excludePath(path: string): boolean {
+  return PATHS_TO_EXCLUDE.some(pathToExclude => {
+    if (path.includes(pathToExclude)) {
+      return true
+    }
   })
```
Comment (knagurski) [inline src/modules/70-pipeline/utils/pipelineReconcileUtils.ts:100]: Not an blocker, but situations like this are perfect for implicit returns
```suggestion
  return PATHS_TO_EXCLUDE.some(pathToExclude => path.includes(pathToExclude))
```

File: src/modules/70-pipeline/utils/__tests__/pipelineReconcileUtils.test.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    )
+    expect(response).toStrictEqual({ hasDifference: false })
+  })
+  test('comparePipelineTemplateAndPipeline - has diff', () => {
+    const response = comparePipelineTemplateAndPipeline(
+      pipelineTemplate2diff as unknown as StageElementConfig,
+      pipelinePipeline2diff
+    )
+    expect(response).toStrictEqual({ hasDifference: true })
+  })
```
Comment (knagurski) [inline src/modules/70-pipeline/utils/__tests__/pipelineReconcileUtils.test.tsx:43]: Try to keep some whitespace between tests to make them easier to visually scan
```suggestion

  test('compareStageTemplateAndStage - has diff', () => {
    const response = compareStageTemplateAndStage(stageTemplate1diff as unknown as StageElementConfig, stagePipeline1)
    expect(response).toStrictEqual({ hasDifference: true })
  })
  
  test('compareStageTemplateAndStage - no diff (exclude spec.environment use case)', () => {
    const response = compareStageTemplateAndStage(stageTemplate2 as unknown as StageElementConfig, stagePipeline2)
    expect(response).toStrictEqual({ hasDifference: false })
  })

  test('comparePipelineTemplateAndPipeline - no diff', () => {
    const response = comparePipelineTemplateAndPipeline(
      pipelineTemplate1 as unknown as StageElementConfig,
      pipelinePipeline1
    )
    expect(response).toStrictEqual({ hasDifference: false })
  })
  
  test('comparePipelineTemplateAndPipeline - has diff', () => {
    const response = comparePipelineTemplateAndPipeline(
      pipelineTemplate2diff as unknown as StageElementConfig,
      pipelinePipeline2diff
    )
    expect(response).toStrictEqual({ hasDifference: true })
  })
```

**Review by knagurski — APPROVED** [review]
No blockers, just some things to keep in mind 👍🏻
